### PR TITLE
Improve test suite

### LINKED ISSUE #97: cleanup tests
**Is your feature request related to a problem? Please describe.**
The tests as they stand are not ideal. They rely on having an active connection to a database and do not have good test coverage, missing many bugs.

**Describe the solution you'd like**
We should take some time to refactor the tests and re-write some more thoughtful tests.

**Additional context**
In issue #90 an in-memory database was added which can be used for tests. Use the top of the authorisation tests for example usage (the test file wasn't fully refactored as we needed to get ts support merged).

### Checklist:

- [x] I have double-checked that the issue does not currently exist.
- [x] I have added the appropriate labels (e.g. bug, enhancement, new feature).
- [x] I have added the **Team 4** project

### PR DESCRIPTION
**Pull request by fishmonger45**
# Description
- Begin working through the tests. Still a bunch of tests to work through and make. Pull atm just includes a draft of auth tests. Still need to reduce duplication of describe blocks and use before/after hooks to reduce code duplication
- Also adds `npm run coverage` for a local coverage report. I don't feel the need to put this in CI/CD because I'm probably the only one who's going to be using it.
- `validate.test.ts`  should be removed when we have proper types in backend

Fixes/resolves #97 

## Type of change

Please delete options that are not relevant.

- [x] **Improvement** (non-breaking change which improves existing functionality)

# Checklist:

Leave blank if not applicable

I have completed these steps when making this pull request:

- [x] I have checked that the PR is from a _forked_ repository
- [x] I have assigned my name to the issue
- [x] I have moved the issue to the **In Progress** column
- [x] I have labelled the PR appropriately
- [ ] I have assigned people responsible to the PR

Before opening the PR for review:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [x] I have documented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have moved the linked issue to the **Review in Progress** column

### CODE REVIEW

**Comment by fishmonger45** [comment]
deploy action has been split to a separate file; no idea if it works or not :cowboy_hat_face:

File: lib/middleware.lib.ts
Comment (mped822) [inline lib/middleware.lib.ts]: If the user is unauthorised shouldn't a 401 unauthorised error be sent instead?

Comment (j-chad) [reply lib/middleware.lib.ts]: in reality, the **401 Unauthorised** status should be called _unauthenticated_. It should be sent if:
- the client sent incorrect login details
- the server could not understand the authentication method
- the client did not send the authentication

**403 Forbidden** should be sent when the server understands the authentication but that user is not _authorised_ to access or change the content.

Comment (fishmonger45) [reply lib/middleware.lib.ts]: I think this was the only way that I could get the test to pass? There's a function that checks the authToken and returns forbidden if it's wrong which is before the unauthorised return of the endpoint function. I don't know if this test even worked originally

Comment (j-chad) [reply lib/middleware.lib.ts]: yeah no worries

**Comment by fishmonger45** [comment]
Please review. I am aware that there are some extensions that are required to be made to the test suite due to new endpoints being introduced. These will be done in a separate PR because we really need some base level tests in the project.

File: package.json
Comment (j-chad) [inline package.json]: we should use NODE_ENV as the environment variable. Also please use cross-env or similar so it's compatible on all platforms

**Review by mped822 — DISMISSED** [review]
lgtm, but i would wait for someone elses review too since i havent had much to do with the backend on this project haha

**Comment by fishmonger45** [comment]
repinging jchad because i suspect it didn't work out of the box because you are on windows? mb. have added crossenv now.

**Comment by j-chad** [comment]
Yeah I am currently on windows 😢. It works well with cross-env, I'm just investigating the issues with the tests that are being skipped

File: config/express.server.config.ts
Comment (j-chad) [inline config/express.server.config.ts]: why was this commented out?

Comment (fishmonger45) [reply config/express.server.config.ts]: couldn't see test results at a glance and scrolling in tmux requires a keyboard command. I'll put it back in.

Comment (j-chad) [reply config/express.server.config.ts]: I'm happy to approve once that is back in

Comment (j-chad) [reply config/express.server.config.ts]: you might want to set the logger verbosity different on tests

**Review by j-chad — APPROVED** [review]
LGTM

**Pull request merged**
